hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/FileChecksumHelper.java

```diff
@@ -316,18 +317,22 @@ FileChecksum makeCompositeCrcResult() throws IOException {
             "Added blockCrc 0x{} for block index {} of size {}",
             Integer.toString(blockCrc, 16), i, block.getBlockSize());
       }
-
-      // NB: In some cases the located blocks have their block size adjusted
-      // explicitly based on the requested length, but not all cases;
-      // these numbers may or may not reflect actual sizes on disk.
-      long reportedLastBlockSize =
-          blockLocations.getLastLocatedBlock().getBlockSize();
-      long consumedLastBlockLength = reportedLastBlockSize;
-      if (length - sumBlockLengths < reportedLastBlockSize) {
-        LOG.warn(
-            "Last block length {} is less than reportedLastBlockSize {}",
-            length - sumBlockLengths, reportedLastBlockSize);
-        consumedLastBlockLength = length - sumBlockLengths;
+      LocatedBlock nextBlock = locatedBlocks.get(i);
+      long consumedLastBlockLength = Math.min(length - sumBlockLengths,
+          nextBlock.getBlockSize());
+      LocatedBlock lastBlock = blockLocations.getLastLocatedBlock();
+      if (nextBlock.equals(lastBlock)) {
```

Comment:
Could you elaborate what this check is? Looking at the test case I assume these few lines distinguish replicated vs striped blocks. Am I right? How about turning them into a helper method that is more readable?

Reply:
Thanks @jojochuang for your comment. 
First, i will explain the goal of UT:

1. Use the same context to create a replicated file and a striped file.
2. Set the conf the use COMPOSITE_CRC
3. Expected  the same checksum result of any length from the replicated file and the striped file via getFileChecksum

Second, i will explain the root cause:

1. blockLocations in line 104, contains a blocks and a lastLocatedBlock
2. the last block in blocks maybe not the same with lastLocatedBlock when the input length is less than file length.
3. so, we cannot always compare with the lastLocatedBlock to get the composer length in line 336.

Reply:
Whether it is a replicated file or striped file, for a block, we will obtain a 4-bytes composer crc, and the actual size corresponding to the crc is very important, because line 341 will use it to compute the composer crc.

Suppose a file has 4 blocks, number block1, block2, block3 and block4 respectively, and the size of each blocks is 10MB, 10MB, 10MB, 7MB. And i use getFilecheck(mockFile, 29MB).  The correct consumedLastBlockLength in line 336 should be 9MB, but the result of the current logic is 7MB which from the last block size of the file.